### PR TITLE
[Merged by Bors] - Parallelize CI for fluvio, fluvio-run, flv-test

### PR DESCRIPTION
**Pull request by nicholastmosher**
This seems to reduce the CI run time (in debug mode, not on `staging`) from about 17-18 mins to about 11 mins

image

### CODE REVIEW

**Comment by nicholastmosher** [comment]
Since some of the job names were changed, it took one execution to refresh the rust-cache. I expect when we run release builds on `staging` it will also take one execution to reset the cache and we will see build time improvements.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,34 +29,53 @@ jobs:
   # build binaries. use release for staging
   # this requires check and test
   build_binaries:
-    name: Build binaries for Target (${{ matrix.rust_target }}) on ${{ matrix.os }}
+    name: Build ${{ matrix.binary }} for ${{ matrix.target }} (${{ matrix.os }})
```
Comment (sehz) [inline .github/workflows/ci.yml:32]: like to keep `rust_target` in order not to get confused with other target

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -29,34 +29,53 @@ jobs:
...
+#          - os: ubuntu-latest
+#            rust: stable
+#            target: armv7-unknown-linux-gnueabihf
+#            binary: fluvio
+#            run: make build-cli
+          - os: macos-latest
+            rust: stable
+            target: x86_64-apple-darwin
+            binary: fluvio
+            run: make build-cli
```
Comment (sehz) [inline .github/workflows/ci.yml:58]: this should only contain matrix configuration not how to perform steps

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -83,57 +102,41 @@ jobs:
...
     strategy:
       matrix:
         os: [ubuntu-latest]
         rust: [stable]
-        rust_target: [x86_64-unknown-linux-gnu]
-        check: [clippy, test, unstable]
+        target: [x86_64-unknown-linux-gnu]
+        check:
+          - name: Rustfmt
+            run: make check-fmt
```
Comment (sehz) [inline .github/workflows/ci.yml:126]: don't add runs here

**Review by sehz — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (nicholastmosher) [reply .github/workflows/ci.yml:58]: The problem is that the old way is really ugly and unweildy, with several disadvantages:

- Each instance of the build job has X steps, one for each binary that might be built, but all except 1 of those are disabled, making the job screen really cluttered and ugly, e.g. here on a target where we only want `fluvio`:

image

- This problem gets worse for every new binary we might want to add to the workflow

When we set up the job this way, it takes much less code and it also displays in the nicest way in the actions UI. I really don't see any disadvantages to doing things this way, it makes it much easier.

Comment (nicholastmosher) [reply .github/workflows/ci.yml:58]: This is opposed to the newer, flatter layout where each job is only instantiated with steps that it actually uses:

image

**Comment by nicholastmosher** [comment]
My goal with structuring the CI is to make it so that, as much as possible, the only thing we should need to do in the future is update the `matrix` section, with minimal changes to the job steps themselves.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (sehz) [reply .github/workflows/ci.yml:58]: that's implementation details.  important point is that `matrix` is configuration.  It should not contain how to perform configuration.

**Comment by nicholastmosher** [comment]
@sehz I took the make commands back out of the matrix

**Review by sehz — APPROVED** [review]
LGTM

**Pull request closed without merging**